app/code/Magento/Payment/Model/MethodList.php

```diff
@@ -39,19 +39,29 @@ class MethodList
      */
     private $paymentMethodInstanceFactory;
 
+    /**
+     * @var array
+     */
+    private $additionalChecks;
+
     /**
      * @param \Magento\Payment\Helper\Data $paymentHelper
-     * @param Checks\SpecificationFactory $specificationFactory
+     * @param Checks\SpecificationFactory  $specificationFactory
+     * @param array                        $additionalChecks
      */
     public function __construct(
         \Magento\Payment\Helper\Data $paymentHelper,
-        \Magento\Payment\Model\Checks\SpecificationFactory $specificationFactory
+        \Magento\Payment\Model\Checks\SpecificationFactory $specificationFactory,
+        array $additionalChecks = []
```

Comment:
It seems a bit awkward to me to inject two dependencies: `$specificationFactory` and its additional configuration in `$additionalChecks`.

Why not pass some `$methodApplicabilityChecks` with default value containing four predefined checks?

Reply:
That was exactly my first thought as well. The factory creates a composite check containing the defined checks, so a simpler implementation would be to just inject different composite checks in different places, and add custom checks to these composite checks where needed.

However, I already discussed this idea with @okorshenko, and he wanted to keep the current whitelisting approach for backwards compatibility reasons. Again, this is how Magento already does it in other places, like [Payment/Block/Form/Container][1].

[1]: https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Payment/Block/Form/Container.php#L93